backend/tournesol/views/subsamples.py

```diff
@@ -0,0 +1,94 @@
+"""
+The sub-samples API return a sub-sample of entities compared by a user.
+
+A sub-sample is a small finite number of compared entities, ranked by the
+individual score computed for the poll's main criterion.
+"""
+import random
+
+import numpy as np
+from drf_spectacular.types import OpenApiTypes
+from drf_spectacular.utils import OpenApiParameter, extend_schema
+from rest_framework import generics
+from rest_framework.permissions import IsAuthenticated
+
+from tournesol.models import ContributorRating
+from tournesol.serializers.subsample import SubSampleSerializer
+from tournesol.views.ratings import ContributorRatingQuerysetMixin
+
+# The number of ranked group used to divide the rated entities.
+DEFAULT_BUCKET_COUNT = 20
+
+
+class SubSamplesQuerysetMixin(ContributorRatingQuerysetMixin):
+    def get_queryset(self):
+        poll = self.poll_from_url
+
+        try:
+            sub_sample_size = int(self.request.query_params["ntile"])
+        except KeyError:
+            sub_sample_size = DEFAULT_BUCKET_COUNT
+
+        # Return a sub-sample of the user's ratings, without annotations so as
+        # not to affect performance.
+        all_ratings = (
+            ContributorRating.objects
+            .filter(
+                poll=poll,
+                user=self.request.user,
+                criteria_scores__criteria=poll.main_criteria
+            )
+            .order_by("-criteria_scores__score")
+            .values_list("entity_id")
+        )
+
+        sub_sample_size = min(sub_sample_size, len(all_ratings))
+        selected = [random.choice(bucket)  # nosec
+                    for bucket in np.array_split(all_ratings, sub_sample_size)]
+
+        # Only the previously selected ratings are retrieved and annotated.
+        sub_sample = (
+            ContributorRating.objects.annotate_n_comparisons()
+            .annotate_collective_score()
+            .annotate_individual_score(poll=poll)
+            .prefetch_related(self.get_prefetch_entity_config())
+            .filter(
+                poll=poll,
+                user=self.request.user,
+                entity_id__in=[rating[0] for rating in selected],
+                criteria_scores__criteria=poll.main_criteria
+            ).order_by("-criteria_scores__score")
+        )
+
+        for idx, item in enumerate(sub_sample):
+            item.bucket = idx + 1
```

Comment:
Any specific reason to start indexing by 1 and not 0 ?

Reply:
Not really. At the beginning I thought we would display the bucket number in the front end, and 1 appeared more user friendly than 0 to identify the first bucket. Now I think using 0 makes everything easier on the back end side.

Reply:
Fixed by https://github.com/tournesol-app/tournesol/pull/1765/commits/310e9fadff85473790ed5ba5f84496e1d655be96